byway/src/main/webapp/recommendations.js

```diff
@@ -0,0 +1,204 @@
+ * Temporary set up to mimic adding several endpoints between legs.
+ */
+function addDestinations(start, end) {
+  destinations.add(start);
+  destinations.add(end);
+  regions.push(...destinations);
+}
+
+/**
+ * Creates a route between two points and loads onto the map.
+ * Finds points along the path to load the regions Array.
+ * Loads recommendations centered around these points.
+ * @param {DirectionsService directionsService} finds directions
+ * @param {DirectionsRenderer directionsRenderer} renders the route
+ * @param {LatLng start} starting point location
+ * @param {LatLng end} ending point location
+ */
+function calcRoute(directionsService, directionsRenderer, start, end) {
+  const request = {
+      origin:  start,
+      destination: end,
+      travelMode: 'DRIVING'
+  };
+  directionsService.route(request, function(response, status) {
+    if (status == 'OK') {
+      directionsRenderer.setDirections(response);
+      findRegions(response);
+      loadRecommendations();
+    } else {
+      alert("Could not calculate route due to: " + status);
+    }
+  });
+}
+
+/**
+ * Goes through steps along route to find the average location
+ * between a step's start and end location. Store these points
+ * in the regions Array.
+ * @param {DirectionsResult directionResult} contains directions
+ * for the route made.
+ */
+function findRegions(directionResult) {
+  const myRoute = directionResult.routes[0].legs[0];
+  for(step of myRoute.steps) {
+    const avgLat = (step.start_location.lat() + step.end_location.lat()) / 2;
+    const avgLng = (step.start_location.lng() + step.end_location.lng()) / 2;
+    const avgLoc = {lat: avgLat, lng: avgLng};
+    if(step.distance.value > MIN_DISTANCE) {
+      regions.push(avgLoc);
+    }
+  }
+}
+
+/**
+ * Set a timeout to delay the browser.
+ * @param {Number delayMs} number of milliseconds
+ */
+function delayPromise(delayMs) {
+  return new Promise(resolve => setTimeout(resolve, delayMs));
+}
+
+/**
+ * Go through a user's interests and search for places
+ * fitting those interests. Search around the regions
+ * previously found using textSearch.
+ */
+async function loadRecommendations() {
+  let isLimited = false;
+  for(interest of interests) {
+    for(region of regions) {
+      await delayPromise(250);
+      const request = {
+        location: region,
+        radius: RADIUS,
+        query: interest
+      }
+      const results = await new Promise((resolve, reject) => {
+        placesService.textSearch(request, (result, status) => {
+          if(status === "OK") {
+            resolve(result);
+          } else {
+            reject(new Error("Status: " + status));
+            if(status === "OVER_QUERY_LIMIT") {
+              isLimited = true;
+            }
+          }
+        });
+      });
+      if(results !== null) {
+        addRecommendations(results);
+      }
+    }
+  }
+  if(isLimited) {
+    alert("Showing limited results");
```

Comment:
Is this message for you the programmer or for an actual user? If it is just helpful for you, it would probably be better to just log this instead of interrupting the page with an alert. If it is a message meant for the user, then it would be better to display a message in the HTML rather than an alert. A small paragraph in red for example to explain the incomplete results.

Reply:
This was meant for the user. I fixed this, but the issue stands when there are multiple "OVER_QUERY_LIMITS" or other status codes s.t. `status !== "OK"`. I'm not sure how to await this part to display the message once, and only after the results are finished.